test/common/stats/thread_local_store_test.cc

```diff
@@ -1816,5 +1923,83 @@ TEST_F(HistogramThreadTest, ScopeOverlap) {
   store_->histogramFromString("histogram_after_shutdown", Histogram::Unit::Unspecified);
 }
 
+size_t getMemoryUsage() {
+#ifndef WIN32
+  std::ifstream status("/proc/self/status", std::ios_base::in);
+  std::string line;
+  const std::string key("VmSize:");
+  while (status.good() && !status.eof()) {
+    std::getline(status, line);
+    const auto pos = line.find(key);
+    if (pos != std::string::npos) {
+      return std::stoull(line.substr(pos + key.size()));
+    }
+  }
+#endif
+  return 0;
+}
+
+// Verify that recording values for histograms that are not sinked does not
+// cause them to grow in memory over time.
+TEST_F(HistogramTest, SinkedHistogramMemoryTest) {
+  StatNamePool pool(store_->symbolTable());
+  std::unique_ptr<Stats::TestUtil::TestSinkPredicates> moved_sink_predicates =
+      std::make_unique<Stats::TestUtil::TestSinkPredicates>();
+  Stats::TestUtil::TestSinkPredicates* sink_predicates = moved_sink_predicates.get();
+  Stats::StatName stat_name = pool.add("h1");
+  sink_predicates->sinkedStatNames().insert(stat_name);
+  store_->setSinkPredicates(std::move(moved_sink_predicates));
+
+  Histogram& h1 = store_->histogramFromStatName(stat_name, Stats::Histogram::Unit::Unspecified);
+  Histogram& h2 = store_->histogramFromString("h2", Stats::Histogram::Unit::Unspecified);
+
+  EXPECT_EQ("h1", h1.name());
+  EXPECT_EQ("h2", h2.name());
+
+  // Create a random number generator to populate histogram values
+  std::random_device random_d;
+  std::mt19937_64 generator(random_d());
+  std::uniform_int_distribution<uint64_t> distribution(0, 1000);
+
+  const size_t num_flushes = 512;
+  const size_t num_updates = 20;
```

Comment:
can we make these numbers much much higher? With such low numbers we might not need to allocate any new pages.

BTW I didn't necessarily expect permanently save a memory test. What I was thinking is that you'd have one that ran for like 10 minutes adding tons of data points to histograms, and look at the process with 'top' while it was running to see if it's increasing in memory.

Reply:
Changed the test to make sure that merge is called for histograms that are not included by sink predicates instead (since that's what we should be testing anyways).